### PR TITLE
Wire notification to sqs

### PR DESCRIPTION
**Pull request by alexduf**
**This should not be merged until** `create-workers-resources` **is merged**

This wires the code that writes to the SQS queues, and also implements a flip/flop filter that decides where to send the notification

### CODE REVIEW

File: notification/app/notification/NotificationApplicationLoader.scala [hunk 1]
```diff
@@ -114,6 +108,55 @@ class NotificationApplicationComponents(context: Context) extends BuiltInCompone
...
+  def withFilter(notificationSender: NotificationSender, invertCondition: Boolean): NotificationSender =
+    new FilteredNotificationSender(notificationSender, topicRegistrationCounter, invertCondition)
+
+  lazy val notificationSenders = List(
+    withFilter(guardianIosNotificationSender, invertCondition = false),
+    withFilter(guardianAndroidNotificationSender, invertCondition = false),
+    withFilter(guardianNewsstandNotificationSender, invertCondition = false),
+    gcmNotificationSender,
+    apnsNotificationSender,
+    newsstandShardNotificationSender,
```
Comment (TBonnin) [inline notification/app/notification/NotificationApplicationLoader.scala]: ok so if I understand this means notification will be pass to all senders? right

File: notification/app/notification/services/FilteredNotificationSender.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+class FilteredNotificationSender(
+  notificationSender: NotificationSender,
+  topicRegistrationCounter: TopicRegistrationCounter,
+  invertCondition: Boolean
+)(implicit ec: ExecutionContext) extends NotificationSender {
+
+  private val logger: Logger = Logger.apply(classOf[FilteredNotificationSender])
+
+  val ALLOWED_TOPICS: Set[TopicType] = Set(TopicTypes.Content, TopicTypes.TagContributor, TopicTypes.TagSeries)
```
Comment (TBonnin) [inline notification/app/notification/services/FilteredNotificationSender.scala]: uppercase?

File: notification/app/notification/services/FilteredNotificationSender.scala [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      }
+    }
+
+    def defaultCountRecovery: PartialFunction[Throwable, PlatformCount] = {
+      case NonFatal(e) if push.notification.topic.exists(_.`type` == TopicTypes.Breaking) =>
+        logger.error(s"Unable to count registration for topics ${push.notification.topic}, falling back to default value", e)
+        PlatformCount(1500000, 750000, 750000, 0)
+      case NonFatal(e) =>
+        logger.error(s"Unable to count registration for topics ${push.notification.topic}, falling back to default value", e)
+        PlatformCount(1, 1, 1, 0)
```
Comment (TBonnin) [inline notification/app/notification/services/FilteredNotificationSender.scala]: There is a risk a notification count fails here, that will make the total count <= 100 and shouldSend will return true when it shouldn't

File: notification/app/notification/services/guardian/ReportTopicRegistrationCounter.scala [hunk 4]
```diff
@@ -8,16 +8,18 @@ import scala.concurrent.duration.DurationLong
...
 )(implicit ec: ExecutionContext) extends TopicRegistrationCounter {
 
   val lruCache: LruCache[PlatformCount] = new LruCache[PlatformCount](200, 1000, 3.days)
 
   override def count(topics: List[Topic]): Future[PlatformCount] = {
     lruCache(topics.toSet) {
       val topicParameters = topics.map(topic => "topics" -> topic.toString)
-      ws.url(registrationCounterUrl)
-        .withQueryStringParameters(topicParameters: _*)
+      val allParams = topicParameters ++ List("api-key" -> apiKey)
```
Comment (TBonnin) [inline notification/app/notification/services/guardian/ReportTopicRegistrationCounter.scala]: can you not append a single tuple: `val allParams = topicParameters :+ ("api-key" -> apiKey)`

File: notification/app/notification/NotificationApplicationLoader.scala (hunk 1 above)
Comment (alexduf) [reply notification/app/notification/NotificationApplicationLoader.scala]: Yes, you're correct, but we filter notifications sent on the SQS for now

**Pull request merged**